src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,201 @@
+    {
+        private readonly IClassificationTypeRegistryService _classificationRegistryService;
+        private readonly IClassificationFormatMap _formatMap;
+
+        public InlineDiagnosticsAdornmentManager(
+            IThreadingContext threadingContext, IWpfTextView textView, IViewTagAggregatorFactoryService tagAggregatorFactoryService,
+            IAsynchronousOperationListener asyncListener, string adornmentLayerName,
+            IClassificationFormatMapService classificationFormatMapService,
+            IClassificationTypeRegistryService classificationTypeRegistryService)
+            : base(threadingContext, textView, tagAggregatorFactoryService, asyncListener, adornmentLayerName)
+        {
+            _classificationRegistryService = classificationTypeRegistryService;
+            _formatMap = classificationFormatMapService.GetClassificationFormatMap(textView);
+            _formatMap.ClassificationFormatMappingChanged += OnClassificationFormatMappingChanged;
+            TextView.ViewportWidthChanged += TextView_ViewportWidthChanged;
+        }
+
+        /// <summary>
+        /// Need to remove the tags if they intersect with the editor view, but only if the option
+        /// to place the tags at the end of the editor is selected.
+        /// </summary>
+        private void TextView_ViewportWidthChanged(object sender, EventArgs e)
+        {
+            if (AdornmentLayer is null)
+            {
+                return;
+            }
+
+            var sourceContainer = TextView.TextBuffer.AsTextContainer();
+            if (sourceContainer is null)
+            {
+                return;
+            }
+
+            if (!Workspace.TryGetWorkspace(sourceContainer, out var workspace))
+            {
+                return;
+            }
+
+            var document = sourceContainer.GetOpenDocumentInCurrentContext();
+            if (document is null)
+            {
+                return;
+            }
+
+            var option = workspace.Options.GetOption(InlineDiagnosticsOptions.Location, document.Project.Language);
+            if (option == InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+            {
+                var normalizedCollectionSpan = new NormalizedSnapshotSpanCollection(TextView.TextViewLines.FormattedSpan);
+                UpdateSpans_CallOnlyOnUIThread(normalizedCollectionSpan, removeOldTags: true);
+            }
+        }
+
+        private void OnClassificationFormatMappingChanged(object sender, EventArgs e)
+        {
+            if (AdornmentLayer is not null)
+            {
+                var elements = AdornmentLayer.Elements;
+                foreach (var element in elements)
+                {
+                    var tag = (InlineDiagnosticsTag)element.Tag;
+                    var classificationType = _classificationRegistryService.GetClassificationType(InlineDiagnosticsTag.TagID + tag.ErrorType);
```

Comment:
the `+` feels allocation heavy.  why can't this be static readonly value that doesn't cause an alloc?

Reply:
How would that work? I need to know the specific tag

Reply:
I don't know if this is an optimization worth investing in. This happens on UI settings configuration (or fonts/colors) right? 

If we had to, the only thing here is ErrorType. We could either do constants for known error types or even just have a dict lookup.

```csharp
var classificationKeys = new Dictionary<string, string>();

foreach (var element in elements)
{
    var tag = (InlineDiagnosticsTag)element.Tag;
    if (!classificationKeys.TryGetValue(tag.ErrorType, out var classificationKey))
    {
        classificationKey = InlineDiagnosticsTag.TagID + tag.ErrorType;
        classificationKeys[tag.ErrorType] = classificationKey;
    }

    var classificationType = _classificationRegistryService.GetClassificationType(classificationKey);
}
````

Reply:
we do it here on every tag: https://github.com/dotnet/roslyn/pull/54083/files#diff-5045c982f077ea1ec41a98a8c76fdd1a8a16ded60d4bcb0ef9cbce5b5cdfffb2R176